### PR TITLE
Runcommand/dynamic

### PR DESCRIPTION
**Pull request by grahamc**
Dynamically define jobs within a jobset by:

1. Enabling them in the hydra.conf:

```xml
<dynamicruncommand>
  enable = 1
</dynamicruncommand>
```

2. Set `enable_dynamic_run_command` to true on the **project record** and the jobset record in the database
3. Create a job in the jobset:

```
{ pkgs, ... }: {
    runCommandHook = {
        recurseForDerivations = true;
        example = pkgs.writeScript "run-me" ''
          #!${pkgs.runtimeShell}
          ${pkgs.jq}/bin/jq . "$HYDRA_JSON"
        '';
    };
}
```

lead-up PRs to this:

* https://github.com/NixOS/hydra/pull/1071
* https://github.com/NixOS/hydra/pull/1072
* https://github.com/NixOS/hydra/pull/1073
* https://github.com/NixOS/hydra/pull/1076

Includes:
* https://github.com/NixOS/hydra/pull/1062

# TODO

- [x] Extend the terraform provider to include dynamic runcommand support flags
    - https://github.com/DeterminateSystems/terraform-provider-hydra/pull/81
- [x] Declarative jobsets should be *rejected* with a useful error if the server or project disable dynamic runcommand and the jobset specifies it as true
    - Semi-complete: we now log that the jobset was rejected, but there's no way (currently) to reject a declarative jobset PUT via the API.
- [x] Extend the API documentation to include the project and jobset dynamic runcommand support flags
- [x] Require the _project_ enable dynamic run command support as well to avoid a declarative jobset from enabling it when it shouldn't be enabled.
- [x] Update runcommand.md to include the requirement that the jobset and project require the feature to be enabled
- [x] The API should reject updates enabling dynamic run command on a jobset if the project does and/or server disable it
- [x] The API should reject updates enabling dynamic run command on a project if server disables it
- [x] Don't let users in the UI enable dynamic runcommands if the server or project or jobset disable it.
- [x] Tell users why they can't enable dynamic runcommands in the UI.
- [x] If the server's config has dynamic runcommands disabled, on the project configuration page say it is not enabled because the server doesn't enable it.
- [x] If the server's config has dynamic runcommands disabled, on the jobset configuration page say it is not enabled because the server doesn't enable it.
- [x] If the project's config has dynamic runcommands disabled, on the jobset configuration page say it is not enabled because the project doesn't enable it.

### CODE REVIEW

File: doc/manual/src/plugins/RunCommand.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+<runcommand>
+  job = myProject:*:*
+  command = cat $HYDRA_JSON > /tmp/hydra-output
+</runcommand>
+```
+
+### Dynamic Commands
+
+Hydra can optionally run RunCommand hooks defined dynamically by the jobset.
+This must be turned on explicitly in the `hydra.conf and per jobset.
```
Comment (Mic92) [inline doc/manual/src/plugins/RunCommand.md:37]:
```suggestion
This must be turned on explicitly in the `hydra.conf` and per jobset.
```

Comment (grahamc) [reply doc/manual/src/plugins/RunCommand.md:37]: Oh nice catch :)

File: src/lib/Hydra/Plugin/GithubStatus.pm [hunk 2]
```diff
@@ -94,7 +94,7 @@ sub common {
                 if (defined $eval->flake) {
                     my $fl = $eval->flake;
                     print STDERR "Flake is $fl\n";
-                    if ($eval->flake =~ m!github:([^/]+)/([^/]+)/([[:xdigit:]]{40})$! or $eval->flake =~ m!git\+ssh://git\@github.com/([^/]+)/([^/]+)\?.*rev=([[:xdigit:]]{40})$!) {
+                    if ($eval->flake =~ m!github:([^/]+)/([^/]+)/([[:xdigit:]]{40})$! or $eval->flake =~ m!git\+ssh://git\@github[^/]+/([^/]+)/([^/]+)\?.*rev=([[:xdigit:]]{40})$!) {
```
Comment (grahamc) [inline src/lib/Hydra/Plugin/GithubStatus.pm:97]: We should break this matcher out in to a separate function, write some tests for it, and upstream it separately. The bit about doCheck should be dropped.

File: src/lib/Hydra/Controller/Jobset.pm [hunk 3]
```diff
@@ -261,13 +261,22 @@ sub updateJobset {
 
     my $checkinterval = int(trim($c->stash->{params}->{checkinterval}));
 
+    my $enable_dynamic_run_command = defined $c->stash->{params}->{enable_dynamic_run_command} ? 1 : 0;
+    if ($enable_dynamic_run_command
+        && !($c->config->{dynamicruncommand}->{enable}
+            && $jobset->project->enable_dynamic_run_command))
+    {
+        badRequest($c, "Dynamic RunCommand is not enabled by the server or the parent project.");
+    }
```
Comment (grahamc) [inline src/lib/Hydra/Controller/Jobset.pm:270]: I'm not sure what the right thing here is, but this pattern of checking has been peppered around. Instead we should probably have a single function which implements "is dynamic run command enabled?" and use that everywhere.

**Comment by cole-h** [comment]
Moved my commits to a separate PR (and also copied over your comments): https://github.com/DeterminateSystems/hydra/pull/4.

**Pull request closed without merging**